File: median-of-two-sorted-arrays/median-of-two-sorted-arrays.go
import "math"
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    if len(nums1) > len(nums2){
            return findMedianSortedArrays(nums2, nums1)
        }
        left, right := 0, len(nums1)
        for left <= right {
            partitionA := left + (right - left)/2
            partitionB := (len(nums1) + len(nums2) + 1)/2 - partitionA

            maxLeftA, maxLeftB := math.MinInt, math.MinInt
            minRightA, minRightB := math.MaxInt, math.MaxInt
            if partitionA > 0 {
                maxLeftA = nums1[partitionA - 1]
            }
            if partitionA < len(nums1){
                minRightA = nums1[partitionA]
            }
            if partitionB > 0 {
                maxLeftB = nums2[partitionB - 1] 
            }
            if partitionB < len(nums2) {
                minRightB = nums2[partitionB]
            }
            //Case 1: Partitions Correct
            if maxLeftA <= minRightB && maxLeftB <= minRightA {
                if (len(nums1) + len(nums2)) % 2 == 0 {
                    return float64((max(maxLeftA, maxLeftB) + min(minRightA, minRightB)))/2.0
                }else{
                    return float64(max(maxLeftA, maxLeftB))
                }
            }else if maxLeftA > minRightB {
                //Case 2: maxLeftA too large for partition A. must make A bigger
                right = partitionA - 1
            }else if maxLeftB > minRightA {
                //Case 3: maxLeftB too small for partition A. need to make A
                left = partitionA + 1
            }
        }
        return 0.0
}

func max(a, b int) int {
    if a >= b {
        return a
    }
    return b
}

func min(a, b int) int {
    if a <= b {
        return a 
    }
    return b 
}

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        //Key intuition here is that the arrays are both sorted
        //Want to do binary search, but need a way to do so without provisioning extra space to maintain runtime requirements. Can partition each array to do this and compare the partitions similar to a min and max heap
        
        //Base case: to avoid overflow error, nums1 must be <= nums2 length
        if(nums1.length > nums2.length) return findMedianSortedArrays(nums2, nums1);

        int left = 0, right = nums1.length; //Will handle
        //[left, right] variant
        while(left <= right) {
            int partitionA = left + (right - left)/2;
            int partitionB = (nums1.length + nums2.length + 1)/2 - partitionA;

            int maxLeftA = partitionA == 0 ? Integer.MIN_VALUE : nums1[partitionA - 1];
            int minRightA = partitionA == nums1.length ? Integer.MAX_VALUE : nums1[partitionA];

            int maxLeftB = partitionB == 0 ? Integer.MIN_VALUE : nums2[partitionB - 1];
            int minRightB = partitionB == nums2.length ? Integer.MAX_VALUE : nums2[partitionB];

            if(maxLeftA <= minRightB && maxLeftB <= minRightA) {
                if((nums1.length + nums2.length) % 2 == 0) return (double) (Math.max(maxLeftA, maxLeftB) + Math.min(minRightA, minRightB))/2;
                else return (double) (Math.max(maxLeftA, maxLeftB));
            }else if(maxLeftA > minRightB) right = partitionA - 1;
            else left = partitionA + 1;
        }
        return -1.0;
    }
}